render/src/umi/mod.rs

```diff
@@ -0,0 +1,307 @@
+        inlines.2.to_inlines()
+    }
+
+    fn fetch_next_line(&mut self, current_line: &mut UmiRow, new_line_index: usize) -> bool {
+        if new_line_index < self.elements.len() {
+            *current_line = self.elements[new_line_index].clone();
+            true
+        } else {
+            false
+        }
+    }
+
+    fn read_row(&mut self, line: usize) -> Block {
+        let mut current_line = self.elements[line].clone();
+        let attributes: HashMap<String, String> =
+            serde_json::from_str(&current_line.attributes).unwrap();
+        match current_line.kind.as_str() {
+            "Heading" => {
+                let heading = Heading {
+                    id: current_line.id.clone(),
+                    level: unimarkup_parser::elements::atomic::HeadingLevel::try_from(
+                        attributes.get("level").unwrap().as_str(),
+                    )
+                    .unwrap(),
+                    content: self.read_inlines(current_line.content.clone()),
+                    attributes: (attributes.get("attributes").cloned()).filter(|s| !s.is_empty()),
+                    start: serde_json::from_str(attributes.get("start").unwrap()).unwrap(),
+                    end: serde_json::from_str(attributes.get("end").unwrap()).unwrap(),
+                };
+                Block::Heading(heading)
+            }
+            "Paragraph" => {
+                let paragraph = Paragraph {
+                    content: self.read_inlines(current_line.content.clone()),
+                };
+                Block::Paragraph(paragraph)
+            }
+            "VerbatimBlock" => {
+                let verbatim = VerbatimBlock {
+                    content: current_line.content.clone(),
+                    data_lang: attributes.get("data_lang").cloned(),
+                    attributes: attributes.get("attributes").cloned(),
+                    implicit_closed: attributes.get("implicit_closed").unwrap().parse().unwrap(),
+                    tick_len: attributes.get("tick_len").unwrap().parse().unwrap(),
+                    start: serde_json::from_str(attributes.get("start").unwrap()).unwrap(),
+                    end: serde_json::from_str(attributes.get("end").unwrap()).unwrap(),
+                };
+                Block::VerbatimBlock(verbatim)
+            }
+            "BulletList" => {
+                let mut bullet_list = BulletList {
+                    entries: vec![],
+                    start: serde_json::from_str(attributes.get("start").unwrap()).unwrap(),
+                    end: serde_json::from_str(attributes.get("end").unwrap()).unwrap(),
+                };
+
+                let bullet_list_depth = current_line.depth;
+                let mut current_line_index = line + 1;
+                self.fetch_next_line(&mut current_line, current_line_index);
+
+                while current_line.depth != bullet_list_depth {
```

Comment:
```suggestion
                while current_line.depth < bullet_list_depth {
```

probably safer, because a nested block may be implicitly closed by a parent.
in this case, the depth would jump 2 or more steps, resulting in an endless loop.